### DISCUSSION TITLE
(FIXED) Build on hpux 10.20

### DISCUSSION THREAD

**Discussion by dcecchino** [XEphem/XEphem#36] [General]
Hello,

I know this is a long shot, but I like running apps still on my old hpux workstation. I am running xephem 3.4, however building 4.1.0 seems to be giving me some issues,  I am erroring out on the htons and https/ssl. Wondering if I am just missing a library, path, or source reference. Any suggestions? 

```
`libz.a' is up to date.
        gcc -L../../libastro -L../../libip -L../../liblilxml -L../../libjpegd -L
../../libpng -L../../libz -L/usr/lib -L/usr/local/lib -L/usr/local/lib/libssl.a 
-L/usr/local/lib/libcrypto.sl.0.9.8 -L/usr/local/lib/libssl.sl.0.9.8  -L-L/usr/d
t/lib -L/usr/lib/Motif1.2 -L/usr/lib/Motif1.2_R6 -L/usr/lib/Motif1.1 -L/usr/lib 
-L/usr/lib/X11R6 -L/usr/lib/X11R5 -L/usr/lib/X11 -L/usr/local/lib -L/usr/local/l
ib/openssl  -o xephem aavso.o            annotmenu.o     broadcast.o     calmenu
.o       closemenu.o     compiler.o      coordsmenu.o    datamenu.o      db.o   
dbmenu.o         earthmap.o      earthmenu.o     fallbacks.o     favmenu.o      
formats.o        fsmenu.o        gallerymenu.o   glance.o        gsc.o          
gscnet.o         helpmenu.o      homeio.o        hznmenu.o       indimenu.o     
imregmenu.o      jpeg2pm.o       jupmenu.o       listmenu.o      mainmenu.o     
marsmenu.o       marsmmenu.o     moonmenu.o      moviemenu.o     msgmenu.o      
netmenu.o        objmenu.o       obslog.o        patchlevel.o    plot_aux.o     
plotmenu.o       preferences.o   progress.o      ps.o            query.o        
rotated.o        satmenu.o       saveres.o       scope.o                 sites.o
 skybinary.o     skyeyep.o       skyfifos.o      skyfiltmenu.o   skyfits.o      
skyhist.o        skyip.o                 skylist.o       skytoolbar.o    skyview
menu.o   solsysmenu.o    splash.o        srchmenu.o      sunmenu.o       time.o 
tips.o           trailmenu.o     uranusmenu.o    ucac.o          usno.o         
versionmenu.o    webdbmenu.o     xe2.o           xe3.o           xephem.o       
xmisc.o -lXt -lX11 -lssl  -lastro -lip -llilxml -ljpegd -lpng -lz -lXm -lXt -lXe
xt -lX11 -lssl -lcrypto -lgcrypt  -L/usr/local/lib -lm -lssl -lcrypto -lgcrypt 
/usr/ccs/bin/ld: Unsatisfied symbols:
   httpsGET (code)
   ssl_recvline (code)
   htons (code)
   ssl_recvlineb (code)
collect2: ld returned 1 exit status
*** Error exit code 1
```

**Comment by brandon-rhodes** [comment]
The routine `httpsGET` is internal to XEphem’s own code:

https://github.com/XEphem/XEphem/blob/main/GUI/xephem/netmenu.c#L507

Could you confirm that your `netmenu.c` contains that function, and that your `netmenu.o` has been freshly rebuilt?

**Reply by dcecchino** [reply]
Just built from source today.

$ ls -l netmenu*
-rw-rw-r--   1 dave       users        35005 Sep 13 12:55 netmenu.c
-rw-rw-r--   1 root       sys          30893 Nov  3 14:00 netmenu.o
$ grep httpsGET netmenu.c
httpsGET (char *host, char *GETcmd, char msg[], XE_SSL_FD *ssl_fd)
$ ls -l netmenu.c
-rw-rw-r--   1 dave       users        35005 Sep 13 12:55 netmenu.c
$ ls -l netmenu.o
-rw-rw-r--   1 root       sys          30893 Nov  3 14:00 netmenu.o

**Reply by brandon-rhodes** [reply]
Then I am stumped. Alas! If the C code defines `httpsGET()`, then I have no idea how the linker could complain `Unsatisfied symbols` about it. You might next want to find the HPUX tools for listing the symbols defined in a `.o` file and confirm that `httpsGET` is a listed symbol in `netmenu.o`. If it is, then you might want to investigate what other problem could arise in the linking phase; I myself don't know of any.

What was the most recent version of XEphem that would compile successfully for you?

**Reply by dcecchino** [reply]
The most recent one that works is 3.4 that compiled. I will take a look at it further, I appreciate you taking the time to help debug it. Thanks!

**Reply by dcecchino** [reply]
So adding the include <netinet/in.h> netmenu.c  resolved the https and ssl errors , so now I just get the 
/usr/ccs/bin/ld: Unnsatisfied symbols: 
htons (code) 
collect2: ld returned 1 exit status 
*** Error exit code 1

**Reply by brandon-rhodes** [reply]
I'm not sure if it's still relevant, but this 18-year-old thread that Google found suggests that same include might be needed in every file that mentions "htons"?

https://comp.sys.hp.hpux.narkive.com/x3FMiYyg/problem-linking-to-htons-etc

**Reply by dcecchino** [reply]
Well, I actually did get it to build adding both includes netinet/in.h and arpa/inet.h to the netmenu.c . But, when I run it it core dumps :-) I'll let you know.

**Reply by dcecchino** [reply]
Looks like /usr/lib/libXm.2 is the issue , core dump showing

#0 0xc34a2950 in _XMGetFocusData+0x78() from /usr/lib/libXm.2

**Reply by brandon-rhodes** [reply]
Alas! I’m not practiced in debugging Motif errors. Can you use a debugger to give you a full stack trace, so we can see which routine was calling `_XMGetFocusData()`? That might let us track down where the error happened in XEphem’s own code.

**Reply by rmathar** [reply]
One way to debug is to add lines like

```
if ( toplevel_w == NULL)
{
    fprintf(stderr,"%d %s NULL pointer toplevel_w\n",__LINE__,__FILE__) ;
    return ;
}
```

before the ``mainrc = XmCreateRowColumn (toplevel_w, "XephemMain", args, n);``  near line 673 in ``xephem.c`` and also   after the 
```
        toplevel_w = XtAppInitialize (&xe_app, myclass, options,
                            XtNumber(options), &argc, argv, NULL, args, n);
```

near line 177, to recompile and to rerun. If XtAppInitialize returns NULL, one may replace that obsolete code by

```
#if 0
        toplevel_w = XtAppInitialize (&xe_app, myclass, options,
                            XtNumber(options), &argc, argv, NULL, args, n);
#else
        /* perhaps wmShellWidgetClass argument here ? RJM 2021-11-11
        */
        toplevel_w = XtOpenApplication (&xe_app, myclass, options,
                            XtNumber(options), &argc, argv, NULL, applicationShellWidgetClass,args, n);
#endif
```
and try again.

**Comment by dcecchino** [comment]
I think this is the best I can do. 
```
# gdb xephem core
HP gdb 3.1.5 for PA-RISC 1.1 or 2.0 (narrow), HP-UX 10.20.
Copyright 1986 - 2001 Free Software Foundation, Inc.
Hewlett-Packard Wildebeest 3.1.5 (based on GDB) is covered by the
GNU General Public License. Type "show copying" to see the conditions to
change it and/or distribute copies. Type "show warranty" for warranty/support.
..
Core was generated by `xephem'.
Program terminated with signal 11, Segmentation fault.

warning: The shared libraries were not privately mapped; setting a
breakpoint in a shared library will not work until you rerun the program.

#0  0xc33eb118 in _XmGetFocusData+0x78 () from /usr/lib/libXm.2
(gdb) bt full
#0  0xc33eb118 in _XmGetFocusData+0x78 () from /usr/lib/libXm.2
No symbol table info available.
#1  0xc33e94e4 in _XmNavigInitialize+0x1c () from /usr/lib/libXm.2
No symbol table info available.
#2  0xc33e4d48 in Initialize+0x220 () from /usr/lib/libXm.2
No symbol table info available.
#3  0xc3640d00 in CallInitialize+0x60 () from /usr/lib/libXt.2
No symbol table info available.
#4  0xc3640cd4 in CallInitialize+0x34 () from /usr/lib/libXt.2
No symbol table info available.
#5  0xc364119c in xtCreate+0x154 () from /usr/lib/libXt.2
No symbol table info available.
#6  0xc3641848 in _XtCreateWidget+0x200 () from /usr/lib/libXt.2
No symbol table info available.
#7  0xc364196c in XtCreateWidget+0x24 () from /usr/lib/libXt.2
No symbol table info available.
#8  0xc33577b0 in create+0x140 () from /usr/lib/libXm.2
No symbol table info available.
#9  0xc3357b58 in XmCreateRowColumn+0x20 () from /usr/lib/libXm.2
No symbol table info available.
#10 0x1ff708 in make_main_window+0xd0 ()
No symbol table info available.
#11 0x1fe4d0 in main+0x278 ()
No symbol table info available
```

**Reply by brandon-rhodes** [reply]
Thanks — that answers our question: it’s in the early `make_main_window()` call. So the window is not even getting set up before the crash. Which at least gives me hope that it wasn’t any tweaks that I myself have made that broke XEphem, as I don’t recall making any tweaks to the main window.

You would need to recompile with symbols and debugging information to learn more about the stack at the moment that it dies. Beyond that it might take actual tracing of pointers and values to figure out what's going wrong.

Another approach would be to download the git repository and bisect: jump forward and back in time looking for the exact commit that broke your ability to launch XEphem.

But, whichever approach is taken, it looks like a full-fledged debugging session would be necessary to figure out what's gone wrong, since I'm not aware of what work has been done on code that early in XEphem's setup code?

**Reply by dcecchino** [reply]
Thanks Brandon, appreciate you looking at it. I will add the debug option as well try git, but not crossing any fingers. Thanks for your help!

**Reply by dcecchino** [reply]
I was able to build it finally, I added two lines to the netmenu.c 

#include <netinet/in.h>
#include <arpa/inet.h>

**Reply by rmathar** [reply]
https://docstore.mik.ua/manuals/hp-ux/en/B2355-60130/byteorder.3N.html
tells us that it might be more portable to add -D_XOPEN_SOURCE_EXTENDED to the compiler instead of using different source files on different systems.
(The better alternative is to convert the build system to autoconfig with a config.h
as usual to deal with different header file combinations...)

**Comment by dcecchino** [comment]
![xephem](https://user-images.githubusercontent.com/17807052/141312257-0a5302e7-3f31-46a1-a2e6-ae28dfb7fd85.jpg)

**Reply by dcecchino** [reply]
Appreciate your help, happy I can keep the usefulness of my old workstation!

**Reply by brandon-rhodes** [reply]
Wonderful! Could you list which suggestions wound up being crucial? That will let us incorporate them into the official source code. Thanks!

**Reply by dcecchino** [reply]
A few things,

In the libz compilation, it was looking for the cc compiler even when I specified gcc in GUI/xephem/Makefile, I had to set zlib/Makefile CC=gcc option

 libz/Makefile 

orion # head Makefile
CLDFLAGS =
CFLAGS = $(CLDFLAGS) -Wall -O2
CC=gcc
LDFLAGS =

I think another important thing in compiling was to make sure that GUI/xephem directory was to perhaps check to make sure that ssl includes are in place and libraries. I did have to manually specify them and also the LIBLNK which was how I got it to compile, also missing the crypt  and crypto libs to make it work for my ancient os. 
GUI/xephem/Makefile
LIBINC = -I../../libastro -I../../libip -I../../liblilxml -I../../libjpegd -I../
../libpng -I../../libz -I/usr/include/netinet/ -I/usr/local/ssl/include
LIBLNK = -L../../libastro -L../../libip -L../../liblilxml -L../../libjpegd -L../
../libpng -L../../libz -L/usr/local/ssl/lib
LIBLIB = -lastro -lip -llilxml -ljpegd -lpng -lz -lnet -lxnet  -lssl -lcrypt -lcrypto

For the actual compilation options

CC = gcc
CLDFLAGS =  -D_XOPEN_SOURCE_EXTENDED
XLIBS = -lXm -lXt -lXext -lX11
LDFLAGS = $(LIBLNK) $(CLDFLAGS) -L$(MOTIFL)
LIBS = $(XLIBS) $(LIBLIB) -lm -lssl -lcrypt  -lcrypto

Not sure why LIBLIB wasn't taking the lcrypt , ssl, and crypto but I guess something odd with my system. But adding it again to LIBS worked.

**Reply by brandon-rhodes** [reply]
All right — I have some time free this week to look at XEphem, so I’d love to get the install instructions updated to include your HP-UX work! It’s amazing to return to this discussion and see the screenshot of XEphem running in CDE. Your hard work paid off!

So:

> In the libz compilation, it was looking for the cc compiler even when I specified gcc in GUI/xephem/Makefile, I had to set zlib/Makefile CC=gcc option

That’s unfortunate! On most Unix systems, the subordinate `make` that’s kicked off by the `libs` target with `cd ../../libz; make` inherits the environment of the parent `make` process which includes `CC`. I just tested, and setting `CC = BLAH` in `GUI/xephem/Makefile` under Linux resulted in the `libz/Makefile` trying to run `BLAH` as its C compiler.

If HP-UX `make` either doesn’t set the environment variable `CC` for its children, or if the subordinate `make` is ignoring it (can we test and determine which is the problem, I wonder?), then I wonder if we should leave a comment in the `# HP-UX` stanza in the main Makefile telling folks to go make the change themselves?

> `LIBINC =` … `LIBLNK =` … `LIBLIB =` …

Instead of having HP-UX folks edit those lines way up at the top of the file, I wonder if we can include those missing elements by putting them in the main “HP-UX” stanza in the main Makefile?

I’ve created a new branch that — in addition to temporarily commenting the Linux variables and un-commenting the HP-UX variables for you — tries adding the include files and libraries you found were missing:

https://github.com/XEphem/XEphem/blob/hpux/GUI/xephem/Makefile#L90

If these became the new officially suggested HP-UX lines (once commented back out), do you think they would work for other users? Is it likely that other folks have put SSL under `/usr/local` as well — is that a usual place for it under HP-UX? Or should a comment instruct folks to customize that?
